### PR TITLE
Add chat api format with standard open ai chat format

### PR DESCRIPTION
**Pull request by elronbandel**

### CODE REVIEW

File: src/unitxt/formats.py [hunk 1]
```diff
@@ -273,19 +312,82 @@ def process(
...
+        target_prefix = self._retrieve_field_and_pop_from_instance(
+            instance=instance, field_name="target_prefix"
+        )
+
+        instance["source"] = (
+            self.tokenizer.apply_chat_template(
+                chat, tokenize=False, add_generation_prompt=True
+            )
+            + target_prefix
         )
```
Comment (OfirArviv) [inline src/unitxt/formats.py:391]: I don't understand why we need this.
Normally we won't be able to have a string "source" string. 
The source should be a dictionary in chat format, no?

Comment (OfirArviv) [reply src/unitxt/formats.py:391]: How do you envision the usage of it?
I envisioned it that the dictionary is in the source, and we just pass this to the model

File: src/unitxt/formats.py [hunk 2]
```diff
@@ -200,43 +206,72 @@ def process(
...
-    _requirements_list = ["transformers"]
+class Message(TypedDict):
+    role: Literal["system", "user", "assistant"]
+    content: Union[str, List[Content]]
 
-    def prepare(self):
-        from transformers import AutoTokenizer
 
-        self.tokenizer = AutoTokenizer.from_pretrained(self.model_name)
+class OpenAIFormat(BaseFormat):
```
Comment (OfirArviv) [inline src/unitxt/formats.py:232]: We need documentation. About the output, the use case and so on. I think I understand what you meant to do, but I'm not sure it generalize well... lets make sure by first you writing what you meant to do

File: src/unitxt/formats.py [hunk 3]
```diff
@@ -200,43 +206,72 @@ def process(
...
-        See more details in https://huggingface.co/docs/transformers/main/en/chat_templating
+Content = Union[TextContent, ImageUrlContent, ImageFileContent]
 
-    """
 
-    model_name: str
-    _requirements_list = ["transformers"]
+class Message(TypedDict):
+    role: Literal["system", "user", "assistant"]
+    content: Union[str, List[Content]]
```
[the remark this answers is not in the document]
Comment (elronbandel) [reply src/unitxt/formats.py:229]: For simple cases it over complicate things. better leave it simple for the 90% of people that do not use image contents.

[the remark this answers is not in the document]
Comment (OfirArviv) [reply src/unitxt/formats.py:229]: it should match the api. Users won't see it. What are apis for HF, OpenAI and etc.?

File: src/unitxt/formats.py (hunk 1 above)
Comment (OfirArviv) [reply src/unitxt/formats.py:441]: Basically the user and unitxt won't know and cannot know what is the format. It only know that the backend somehow handle this

Comment (elronbandel) [reply src/unitxt/formats.py:441]: This is just for backward compatibility (for pepole who's using unitxt to get data with hf chat template in a textual format).

File: src/unitxt/formats.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (elronbandel) [reply src/unitxt/formats.py:229]: Pepole do look at what their data look like and they do want to see over complication. What are we losing by giving pepole output that is easier to understand and that is in a format they are mostly familiar with?

File: src/unitxt/formats.py (hunk 2 above)
Comment (elronbandel) [reply src/unitxt/formats.py:232]: done

File: src/unitxt/formats.py (hunk 1 above)
Comment (OfirArviv) [reply src/unitxt/formats.py:440]: Separate it to a different mechanism imo. We don't need to mix stuff that are not related from design pov

File: src/unitxt/formats.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (OfirArviv) [reply src/unitxt/formats.py:228]: it should reflect exactly what going into the model. That the most simple thing

[the remark this answers is not in the document]
Comment (OfirArviv) [reply src/unitxt/formats.py:228]: if there are multiple options that these api can recieve thats another matter.

File: src/unitxt/formats.py (hunk 2 above)
Comment (OfirArviv) [reply src/unitxt/formats.py:231]: write what this format is. Give an example. If there are multiple formats list them

**Comment by yoavkatz** [comment]
@elronbandel - I think we should make this self contained  
1. Add a HF inference engine (that uses the chat API)
2. Add en example code that calls this API.
3. Add it to catalog.

**Comment by OfirArviv** [comment]
> * Add a HF inference engine (that uses the chat API)
> * Add en example code that calls this API both as text and with image.

We should wait for Pawel pr for this. The changes in the inference engine is small. No need to duplicate it

**Pull request merged**